static/tutorial/exception-safety.md

```diff
@@ -0,0 +1,341 @@
+# Exception Safety
+
+This tutorial covers a relatively broad and previously underdocumented
+topic in the Haskell ecosystem. In particular, goals of this tutorial
+are:
+
+* Explain the different types of exceptions
+* Explain how exceptions can be generated
+* Explain how interruptible actions work
+    * More to the point: interruptible actions introduce a concept of
+      trading off between *correctness* and *deadlock prevention*
+* Define requirements for proper exception safety in
+    * Normal code
+    * Resource acquisition code
+    * Resource cleanup code
+* Give guidelines on how to write exception safe code in general
+* Point out some existing violations of these rules in the Haskell
+  ecosystem
+
+Many of the ideas behind this tutorial originated with the
+[safe-exceptions](/package/safe-exceptions) package, and as such it is
+strongly recommended that you work with that package to get maximum
+correctness and ease-of-use.
+
+Like the safe-exceptions package, this tutorial takes runtime
+exceptions in Haskell as a fact, not a point for discussion, as this
+is the reality of the library ecosystem today. No guidelines are given
+here for when you yourself should generate runtime exceptions versus
+using something like `IO (Either MyError a)` or `ExceptT`. That
+discussion is left for a different time.
+
+## Types of exceptions
+
+We're going to follow the safe-exceptions package's definitions for
+exception types. It is strongly recommended that you read
+[the safe-exceptions tutorials](/package/safe-exceptions) first. As a
+quick summary:
+
+* While there is a difference in the type hierarchy for synchronous
+  and asynchronous exceptions (via the `SomeAsyncException` type),
+  this distinction is not enforced by the `Control.Exception` module
+  (e.g., you can synchronously throw via `throwIO` an
+  asynchronously-typed exception).
+* Despite this limitation, we assume that we can fully distinguish
```

Comment:
The type level distinction always existed but wasn't enforced. As it happens, it was always very uncommon to send a synchronously typed exception via throwTo, and vice versa. We're not the first people to try and use the type level distinction, we're just the first to codify an entire replacement module that way.